core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,178 @@
+    this.beforeExpiration = beforeExpiration;
+    this.afterExpiration = ops.refresh();
+  }
+
+  @Override
+  public void cleanFiles() {
+    // Identify all of the manifest lists to retain
+    Set<String> manifestListsBeforeExpiration =
+        ReachableFileUtil.manifestListLocations(beforeExpiration);
+
+    // Identify all of the manifest lists that currently exist
+    Set<String> manifestListsAfterExpiration =
+        ReachableFileUtil.manifestListLocations(afterExpiration);
+
+    Set<String> manifestListsToDelete =
+        Sets.difference(manifestListsBeforeExpiration, manifestListsAfterExpiration);
+    Map<String, ManifestFile> reachableManifestsAfterExpiration =
+        computeManifestsForManifestLists(manifestListsAfterExpiration);
+
+    // Compute the manifests to delete based on the current manifest lists to delete and the current
+    // reachable manifests
+    List<ManifestFile> manifestsToDelete =
+        computeManifestsToDelete(manifestListsToDelete, reachableManifestsAfterExpiration);
+
+    List<ManifestFile> currentManifests =
+        Lists.newArrayList(reachableManifestsAfterExpiration.values());
+
+    Set<String> filesToDelete = findFilesToDelete(manifestsToDelete, currentManifests);
+    Tasks.foreach(filesToDelete)
+        .executeWith(deleteExecutorService)
+        .retry(3)
+        .stopRetryOn(NotFoundException.class)
+        .suppressFailureWhenFinished()
+        .onFailure((file, exc) -> LOG.warn("Delete failed for data file: {}", file, exc))
+        .run(deleteFunc::accept);
+
+    Set<String> manifestPathsToDelete =
+        manifestsToDelete.stream().map(ManifestFile::path).collect(Collectors.toSet());
+
+    LOG.warn("Manifests to delete: {}", Joiner.on(", ").join(manifestsToDelete));
+    LOG.warn("Manifests Lists to delete: {}", Joiner.on(", ").join(manifestListsToDelete));
+    deleteMetadataFiles(
+        manifestPathsToDelete,
+        (path, exc) -> LOG.warn("Delete failed for {}: {}", "manifest", path, exc));
+    deleteMetadataFiles(
+        manifestListsToDelete,
+        (path, exc) -> LOG.warn("Delete failed for {}: {}", "manifest list", path, exc));
+  }
+
+  private List<ManifestFile> computeManifestsToDelete(
+      Set<String> manifestListsToDelete, Map<String, ManifestFile> reachableFilesAfterExpiration) {
+    Map<String, ManifestFile> manifestCandidatesToDelete =
+        computeManifestsForManifestLists(manifestListsToDelete);
+    List<ManifestFile> manifestFilesToDelete = Lists.newArrayList();
+    for (Map.Entry<String, ManifestFile> candidateManifest :
+        manifestCandidatesToDelete.entrySet()) {
+      if (!reachableFilesAfterExpiration.containsKey(candidateManifest.getKey())) {
+        manifestFilesToDelete.add(candidateManifest.getValue());
+      }
+    }
+    return manifestFilesToDelete;
+  }
+
+  private Map<String, ManifestFile> computeManifestsForManifestLists(Set<String> manifestLists) {
```

Comment:
For determining the manifest lists I'm using ReachableFileUtil which already handles the validation that the manifest list may not exist for a snapshot. So it's the same from a correctness standpoint. Or are you referring to if it's cleaner to use the snapshots?

Reply:
Oh actually I see what you mean, in some cases on the snapshot itself instead of the manfiest list location we have a list of the actual manifest file locations. Will update to use snapshots then.